sgkit/stats/decomposition.py

```diff
@@ -0,0 +1,169 @@
+
+from ..typing import ArrayLike
+
+
+# https://github.com/dask/dask-ml/blob/b94c587abae3f5667eff131b0616ad8f91966e7f/dask_ml/_utils.py#L15
+# Grabbing this from dask-ml to avoid declaring a dependency on dask-ml
+def draw_seed(random_state, low, high=None):  # type: ignore
+    return random_state.randint(low, high)
+
+
+class GenotypePCA(sklearn.decomposition.PCA):  # type: ignore
+    """
+
+    Parameters
+    ----------
+    copy : boolean, optional, default True
+        ignored
+    ploidy : int, optional, default 2
+        The n_ploidy of the samples. Assumed to be 2 for diploid samples
+    n_components: int, optional, default 10
+        The estimated number of components.
+
+    Attributes
+    ----------
+    mean_ : ndarray or None, shape (n_variants, 1)
+        The mean value for each feature in the training set.
+    std_ : ndarray or None, shape (n_variants, 1)
+        scaling factor
+
+    Differences from scikit-allel
+    ----------
+    * The scalers have been separated out from the PCAs to conform with
+    SKLearn Pipelines - https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html
+
+    * Uses Dask under the hood instead of numpy
+    * svd_solver : 'randomized' uses ``dask.linalg.svd_compressed``
+      'full' uses ``dask.linalg.svd``, 'arpack' is not valid.
+    * iterated_power : defaults to ``0``, the default for
+      ``dask.linalg.svd_compressed``.ad of numpy
+
+    Examples
+    --------
+    >>> from sgkit.stats.preprocessing import PattersonScaler
+    >>> from sgkit.stats.decomposition import GenotypePCA
+    >>> import dask.array as da
+
+    >>> # Let's generate some random diploid genotype data
+    >>> # With 30000 variants and 67 samples
+    >>> n_variants = 30000
+    >>> n_samples = 67
+    >>> genotypes = da.random.choice(3, n_variants * n_samples)
+    >>> genotypes = genotypes.reshape(n_variants, n_samples)
+
+    >>> scaler = PattersonScaler()
+    >>> scaled_genotypes = scaler.fit_transform(genotypes)
+    >>> # If you want to deal with the scaled values directly
+    >>> # scaled_genotypes.compute()
+    >>> # Or you can put the scaled_genotypes directly into the PCA
+    >>> pca = GenotypePCA()
+    >>> transformed = pca.fit_transform(scaled_genotypes)
+
+    >>> # Use SKLearn Pipelines
+    >>> # https://github.com/pystatgen/sgkit/issues/95#issuecomment-672879865
+    >>> from sklearn.pipeline import Pipeline
+    >>> est = Pipeline([ \
+        ('scaler', PattersonScaler()), \
+        ('pca', GenotypePCA(n_components=2)) \
+        ])
+    >>> pcs = est.fit_transform(genotypes)
+    >>> # `est` would also contain loadings + explained variance
+    >>> # `scaler` would contain the MAF and binomial variance values needed for out-of-sample projection
+    >>>   # Out-of-sample projection
+    >>> pcs_oos = est.transform(genotypes)
+    """
+
+    def __init__(
+        self,
+        n_components: int = 10,
+        copy: bool = True,
```

Comment:
The dask arrays aren't overwritten. I think because they are calculated within the class? 

I mostly just tried to stick with the choices dask-ml made.